File: asteroidz-web/tools/native-run.js
// Run the AsteroidZ wasm natively under Node's WASI - no browser, no webview,
// no canvas. Every env.* host function the web runtime provides is auto-
// stubbed from the module's own import table; Swift print() reaches stdout
// through WASI fd_write, so game logs appear directly in the terminal.
//   node tools/native-run.js <wasm> [seconds]
const { readFileSync } = require('node:fs');
const { WASI } = require('node:wasi');

const wasmPath = process.argv[2] || 'web/asteroidz.wasm';
const seconds = Number(process.argv[3] || 5);

(async () => {
  const bytes = readFileSync(wasmPath);
  const module = await WebAssembly.compile(bytes);

  const wasi = new WASI({ version: 'preview1', args: [], env: {} });

  // Auto-stub the web runtime's host surface: void for most, 0 for handles.
  const env = {};
  let gfxCalls = 0;
  for (const imp of WebAssembly.Module.imports(module)) {
    if (imp.module !== 'env' || imp.kind !== 'function') continue;
    const name = imp.name;
    env[name] = name.startsWith('gfx_') ? (() => { gfxCalls++; return 0; })
                                        : (() => 0);
  }

  const instance = await WebAssembly.instantiate(module, {
    wasi_snapshot_preview1: wasi.wasiImport,
    env,
  });
  wasi.initialize(instance);

  instance.exports.boot();
  const frames = Math.round(seconds * 60);
  for (let i = 0; i < frames; i++) instance.exports.frame(16.67);

  console.log(`\n--- native WASI run complete: ${frames} frames (${seconds}s of game time), ` +
              `${gfxCalls.toLocaleString()} draw calls issued, no browser involved ---`);
})().catch(e => { console.error('FAILED:', e.message); process.exit(1); });

File: asteroidz-web/tools/native-run.py
# Run the AsteroidZ wasm under wasmtime (WASI Preview 1) - no browser, no JS,
# no webview. The web runtime's env.* surface is auto-stubbed from the
# module's import table; Swift print() reaches stdout via WASI fd_write.
#   python3 tools/native-run.py <wasm> [seconds]
import sys
from wasmtime import Store, Module, Linker, Engine, WasiConfig, FuncType, ValType

wasm = sys.argv[1] if len(sys.argv) > 1 else "web/asteroidz.wasm"
seconds = float(sys.argv[2]) if len(sys.argv) > 2 else 5.0

engine = Engine()
store = Store(engine)
wasi = WasiConfig()
wasi.inherit_stdout()
wasi.inherit_stderr()
store.set_wasi(wasi)

module = Module.from_file(engine, wasm)
linker = Linker(engine)
linker.define_wasi()

calls = {"gfx": 0}
for imp in module.imports:
    if imp.module != "env":
        continue
    ft = imp.type
    name = imp.name
    def make(name, ft):
        zeros = [0] * len(ft.results)
        def stub(*args):
            if name.startswith("gfx_"):
                calls["gfx"] += 1
            if len(zeros) == 0:
                return None
            return zeros[0] if len(zeros) == 1 else zeros
        return stub
    linker.define_func("env", name, ft, make(name, ft))

instance = linker.instantiate(store, module)
exports = instance.exports(store)
exports["_initialize"](store)
exports["boot"](store)
frames = round(seconds * 60)
for _ in range(frames):
    exports["frame"](store, 16.67)

print(f"\n--- wasmtime run complete: {frames} frames ({seconds:g}s), "
      f"{calls['gfx']:,} draw calls, zero JavaScript involved ---")
